### DISCUSSION TITLE
Automation loop [Object object] error

### DISCUSSION THREAD

**Discussion by hansblaauw64** [Budibase/budibase#12360] [Help]
I'm returning JSON from an API call with the below results (list of results). When I try to parse it in a loop it gives me [Object object] error.

I need to go over every result in the JSON and return the DATA MACHINE part.

When using binding value {{ steps.2.response}} I cannot iterate and i'm getting the above error.

Any suggestions?

{
  "response": [
    {
      "partition": 0,
      "offset": 228,
      "dataType": "demo",
      "data": {
        "correlationId": "5-1370476776",
        "version": 1,
        "timestamp": "2023-11-03T15:20:58+02:00",
        "receivingTime": "2023-11-03T15:21:18.743+02:00",
        "machine": {
          "companyUniqueId": "005-942",
          "uniqueId": "005-88986",
          "externalId": "V1117",
          "pid": "2202603243",
          "deviceType": "TE",
          "locationExternalId": "Test_Location",
          "locationName": "Test",
          "name": "Test"
        },
        "uniqueId": "005-1370476776",
        "transactionId": "xxxxxxxxxxxxxxxxxxxxxxxxxx",
        "column": 49,
        "quantity": 1,
        "amount": 0.01,
        "currency": "EUR",
        "taxRate": 0,
        "product": null,
        "recipe": null,
        "voucher": null,
        "company": {
          "uniqueId": "005-942",
          "name": "Demo",
          "externalId": "",
          "address": "Address",
          "vatNumber": "",
          "vatPayer": null
        },
        "attributes": {}
      }
    },

**Comment by sdsys-ch** [comment]
Hi @hansblaauw64 
Is the external api endpoint returning valid json?
Your given example isn't. 
If the above is just an extract, try posting a valid subset of the returned data. `response` should be iterable since it points to an array structure of dicts (that would need to be handled individually to get to the `machine`-key) but can't say for sure with the data you posted.

**Comment by hansblaauw64** [comment]
It is returning valid JSON, see below. My problem is how to get to the data section in a loop in an automation. stuff like {{ steps.2.response.data}} and {{ loop.currentItem.machine.name }} didn't work.
![looping](https://github.com/Budibase/budibase/assets/39296077/cd0ebf24-f9d7-4538-a181-2cfcf2228c00)

The response part is added by Budibase in the automation.

    {
    "response": [
        {
            "partition": 0,
            "record": 332,
            "dataType": "items",
            "data": {
                "correlationId": "5-1374128696",
                "version": 1,
                "timestamp": "2023-11-06T18:45:28+02:00",
                "receivingTime": "2023-11-06T18:45:44.58+02:00",
                "machine": {
                    "companyUniqueId": "005-942",
                    "uniqueId": "005-88986",
                    "externalId": "test",
                    "pid": "2202603243",
                    "deviceType": "TE",
                    "locationExternalId": "Test_Location",
                    "locationName": "Test location test",
                    "name": "Test machine 2"
                },
                "uniqueId": "005-1374128696",
                "transactionId": "2202603243-268449034",
                "column": 53,
                "quantity": 1,
                "amount": 0.01,
                "currency": "EUR",
                "taxRate": 0,
                "inTestMode": false,
                "failedVend": false,
                "product": {
                    "uniqueId": "151760",
                    "externalId": "8435185944009",
                    "name": "Product",
                    "depositRefundAmount": 0,
                    "barcode": {
                        "barcode": "8435185944009",
                        "barcode1": "",
                        "barcode2": "",
                        "barcode3": "",
                        "barcode4": ""
                    }
                },
                "recipe": null,
                "voucher": null,
                "company": {
                    "uniqueId": "005-942",
                    "name": "Test",
                    "externalId": "",
                    "address": "Address",
                    "vatNumber": "",
                    "vatPayer": null
                },
                "attributes": {}
            }
        },
        {
            "partition": 0,
            "record": 435,
            "dataType": "items",
            "data": {
                "correlationId": "5-1377811024",
                "version": 1,
                "timestamp": "2023-11-08T14:32:39+02:00",
                "receivingTime": "2023-11-08T14:32:56.145+02:00",
                "machine": {
                    "companyUniqueId": "005-942",
                    "uniqueId": "005-88986",
                    "externalId": "dsdsdsdsd",
                    "pid": "2202603243",
                    "deviceType": "TE",
                    "locationExternalId": "Test_Location",
                    "locationName": "Test location test",
                    "name": "Test machine 2"
                },
                "uniqueId": "005-1377811024",
                "transactionId": "2202603243-268449137",
                "column": 59,
                "quantity": 1,
                "amount": 0.16,
                "currency": "EUR",
                "taxRate": 0,
                "inTestMode": false,
                "failedVend": false,
                "product": null,
                "recipe": null,
                "voucher": null,
                "company": {
                    "uniqueId": "005-942",
                    "name": "Test",
                    "externalId": "",
                    "address": "Address",
                    "vatNumber": "",
                    "vatPayer": null
                },
                "attributes": {}
            }
        }
    ],
    "info": {
        "pagination": {
            "cursor": null
        },
        "headers": {
            "date": "Sat, 11 Nov 2023 08:15:47 GMT",
            "content-type": "application/json",
            "content-length": "102501",
            "connection": "close",
        },
        "code": 200,
        "size": "100KB",
        "time": "4203ms"
    },
    "success": true
}

**Comment by sdsys-ch** [comment]
Hi @hansblaauw64 

Just tested this via a webhook sending your above json data (and fixing line 108) and using 

`{{ trigger.response }}` as binding for the array input type, and
` {{ loop.currentItem.data.machine.externalId }}` & `{{ loop.currentItem.data.machine.name }}` for the table entries:

![image](https://github.com/Budibase/budibase/assets/100968265/fce09efe-326c-48ff-8fde-1c2b01bb2756)

This works as expected:

![image](https://github.com/Budibase/budibase/assets/100968265/951a839e-78c1-4b97-a41d-075475b971fe)

and creates the related tables rows:

![image](https://github.com/Budibase/budibase/assets/100968265/16919c8f-e51a-4437-92c4-31e9fb0ded0c)

In your example you're trying to loop over the json **dictionary** `data` via the input loop binding instead of looping over the json **array** `response`. 

In the second step, when creating the row, it loops over the entries of said array. Simply select the relevant dict-entries via their key path in order to fill your table, e.g. `{{ loop.currentItem.data.machine.uniqueId }}`, `{{ loop.currentItem.data.correlationId }}` etc.

Since your not getting your data from a webhook, simply use `{{ steps.2.response }}` as array loop binding, making sure that's where the data's at. This should do it.

**Comment by hansblaauw64** [comment]
Hi, this is what I had already had and it does not work. It does not iterate over the results. I'm getting stuff from an API with Budibase and then Budibase adds the response object. Using {{ steps.2.response }} does not work, I already did this.

Assigning values to the table is something I already had too and that part is clear.

When I use the {{steps.2.response}} as binding i'm getting the [object Object error].

**Reply by sdsys-ch** [reply]
My bad. Can you post a JSON stringified version of the `{{ steps.2.response }}` object (get it via JS helpers `stringify`-function) here? And maybe a screenshot of your actual loop settings as well as any input/output of a failed loop test? That might help. I'm not on my desktop now but will look at it later today.

EDIT: Came back to see if there was an update. Question: are you sure you're using the correct input binding in your automation? In your above automation screenshot you are using `{{ steps.2.response.data }}`. Might be from an old test but could be worthwhile checking again.

**Comment by hansblaauw64** [comment]
After doing so more tests it seems it works when I limit the iterations in the loop. We run self hosted and I think we hit the max iterations limit. If I run 30 iterations all is fine.

**Reply by sdsys-ch** [reply]
Well, glad you got a little further. Still seems weird. I think the hard limit is set at 200 by default? It would be interesting to see what happens if you set AUTOMATION_MAX_ITERATIONS...
User set limits also get reported as error in the status field of the response when reached. Something similar should apply to the hard limit, when reached. The error you received points to a wrong binding, I believe. Otherwise it could be a bug... Sorry I wasn't of more help. Hope you solve it.

**Comment by hansblaauw64** [comment]
Must be a bug, it iterates 49 times without error and then suddenly the object error. Json is correct. But despite error it handles all json and adds rows. The error is annoying, it is an alert which stays in screen.
